compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -881,6 +869,47 @@ fn convert_path_expr<'a, 'tcx>(
     }
 }
 
+fn convert_static_ref<'tcx>(cx: &mut Cx<'_, 'tcx>, id: DefId, ty: Ty<'tcx>) -> ExprKind<'tcx> {
+    if cx.tcx.is_thread_local_static(id) {
+        ExprKind::ThreadLocalRef(id, ty)
+    } else {
+        let ptr = cx.tcx.create_static_alloc(id);
+        ExprKind::StaticRef {
+            literal: ty::Const::from_scalar(cx.tcx, Scalar::Ptr(ptr.into()), ty),
+            def_id: id,
+        }
+    }
+}
+
+fn convert_addr_of_expr<'tcx>(
+    cx: &mut Cx<'_, 'tcx>,
+    kind: hir::BorrowKind,
+    mutability: hir::Mutability,
+    arg: &'tcx hir::Expr<'tcx>,
+) -> ExprKind<'tcx> {
+    // Fast path so that taking a reference to a static doesn't end up
+    // as `&*&STATIC` but just `&STATIC`
+    if let hir::ExprKind::Path(qpath) = &arg.kind {
+        let res = cx.typeck_results().qpath_res(qpath, arg.hir_id);
+        if let Res::Def(DefKind::Static, id) = res {
+            let mut tm = cx.tcx.static_ptr_ty(id).builtin_deref(true).unwrap();
+            tm.mutbl = mutability;
```

Comment:
This looks like shoehorning the existing `static_ptr_ty` helper into something else. Is that helper still used elsewhere, given the new shape of things?

Reply:
yes, it's still used further up. I didn't want to repeat its code, but I could split out the part that obtains the static's type into a separate function

Reply:
In particular, that function contains some subtle logic if I recall correctly (for using raw ptrs vs references), and now we overlay our own logic on top of this here... given how critical this code is, I'd prefer that to be cleaner, even if that makes the code longer.

Also it would be good to have some comments explaining why all of this here crucially has to be exactly the way it is.

Reply:
> yes, it's still used further up.

Ah, I just saw that.  But that further up should now not be reachable any more for refs-to-static, right, just for when it is used as a value? If so, please add a comment explaining that. Also, doesn't that mean we can always use a shared reference, as we always just read? Well and I guess a raw pointer for unsafe-to-read things to make sure we do not sidestep the unsafety checker.

I feel like all the subtle logic that determines the type of the reference we use should be in one place.